### ISSUE TITLE
[Enhancement]: Allow multiple instances of Mini Cart block

### ISSUE THREAD

**Issue by webmandesign** [woothemes/woocommerce#46166]
### Describe the solution you'd like

The Mini Cart block can currently be displayed/used only once.

However, in my block theme I am building a desktop and mobile site header template part and would like to use the Mini Cart block in each of these headers.

I've tried to enable multiple instances for the Mini Cart block via `blocks.registerBlockType` JavaScript filter hook, which works, but it  also produces multiple Mini Cart overlays (the `.wc-block-components-drawer__screen-overlay`) when user adds a product to cart. This is obviously not ideal as user has to close 2 Mini Cart overlays then.

Can you allow multiple Mini Cart block instances that open just one Mini Cart overlay, please?

**Comment by Marc-pi** [comment]
yep, so do it (example in conjunction with Nick Diego's block visibility plugin and the same target aka mobile menu)

**Comment by senadir** [comment]
Would this be 2 Mini Cart triggers? or are you building two different drawers all together?

**Comment by Marc-pi** [comment]
The same mini cart in my case
My use case : 
- on desktop, the cart icon is within the navigation, as the latest entry, after the user account account

![image](https://github.com/woocommerce/woocommerce/assets/919332/289cb4f6-d459-47d1-b9ae-27dd965fac89)


- on mobile, we have the burger menu : the cart icon must be displayed outside the nav component, and first.

![image](https://github.com/woocommerce/woocommerce/assets/919332/0cfc3768-fd32-49c4-a821-645d134fa1da)

- currently, and **only** if you edit the header template part, we get the message about unicity, strange !
   - this message is not displayed if you edit the page template directly from the page itself 
![image](https://github.com/woocommerce/woocommerce/assets/919332/abeaa233-ca18-4bb3-86b2-0aa996a3b8af)

**Comment by webmandesign** [comment]
@senadir 

I'm not really sure I understand your questions 100%, but will try to answer.

> Would this be 2 Mini Cart triggers? 

I think so. Basically I am displaying 2 Mini Cart blocks that should trigger just 1 drawer.

It would be nice to be able to use any number of Mini Cart blocks on a page and all of these blocks should trigger a single drawer instead producing drawer for each Mini Cart block instance.

> or are you building two different drawers all together?

No, I don't want any multiple drawers. That's the issue here for me personally: multiple drawers are produced so user has to close them all. It looks really strange now as you close the first drawer and there is still a drawer displayed. User with not dev experience doesn't really know what is happening and the website seems broken.

There should be **just 1 drawer** displayed regardless of the number of Mini Cart blocks used.

**Comment by wavvves** [comment]
Hi @webmandesign, 

I have been trying this out, and it seems you can have multiple co-existing triggers (and drawer is always in sync, closing it and opening works from all), just not in the same section eg one in the header and one in footer is possible.

Can this be used to circumvent your limitation? Perhaps creating a new template pattern containing the mini-cart that you can insert again in the header pattern?

**Comment by opr** [comment]
Hi @webmandesign please could you let us know if the above worked for you? If it's a valid solution then I suggest we close the issue. Thanks!

**Comment by webmandesign** [comment]
Hi @wavvves & @opr,

Sorry for delay in my answer.

Please note that the issue still persist. Please understand the issue is related to **when adding a product to cart** - the "Open drawer when adding" option is enabled for the Mini Cart block.

When I add the Mini Cart to footer template part, the issue still persist. It does not matter where the Mini Cart is displayed. If I add 3 Mini Carts on the page **and I add a product to cart**, 3 drawers are open stacked and I have to close all 3 of them.

When I only open Mini Cart by clicking on the cart icon, only 1 drawer open, which is OK. But that's not the issue. The issue occurs **when adding a product to cart** and there are multiple Mini Cart blocks on the page as described.

**Comment by webmandesign** [comment]
Here I've create a short video of the issue:

https://github.com/woocommerce/woocommerce/assets/2456735/afa4579b-8e66-44e1-9997-5738ba95b51b

- **00:00** - Preview of how Header template part is set: you can see 2 header containers - one for desktop and one for mobile - with Mini Cart in each of them. Mini Cart is set with "Open drawer when adding" option (which is default, if I'm not mistaken?).
- **00:27** - I am adding Mini Cart block to Footer template part, as suggested by @wavvves.
- **00:49** - Previewing Mini Cart block functionality on Shop page: clicking the Mini Cart icon (any of them) works just fine as it opens just 1 overlay drawer.
- **01:12** - Previewing the issue: trying to add a product to cart opens Mini Cart overlay drawer 3 times. You can see it even on opacity of the overlay as it gets darker and now I have to close 3 overlays: for 2 Mini Carts in header and 1 in footer.

Hope this helps visualizing the issue.

**Comment by opr** [comment]
@webmandesign thanks a lot, this is really helpful! I will add it to our backlog to be addressed.

**Comment by pmcpinto** [comment]
I'll remove the high-priority label, as we don't have the bandwidth to implement this in cycle 4. However, it should be considered for the following cycle, and it would be helpful to estimate the effort required to move forward with this change. Thanks.

**Comment by samueljseay** [comment]
This is definitely a bug, but, I just wanted to mention there is a workaround for @webmandesign, if you ensure that only one of the mini cart blocks present on the page has the "Open drawer when adding" behaviour, it will still open the cart when adding an item but you won't get multiple drawers stacked.

**Comment by webmandesign** [comment]
@samueljseay That's a great advice, thank you! I've disabled the "Open drawer when adding" behavior on both Mini Cart blocks, but now that you mentioned it, I can actually enable it for the desktop one.

**Comment by pierorocca** [comment]
@samueljseay has the bug portion of this issue been separated out of this enhancement request? 

I'd like to get some design feedback on the original issue of mobile vs. desktop header differences and/or responsiveness. I'm not convinced that the workaround solution addresses the problem in the most optimal or performant way, and it likely complicates store editing.

In my test site below, the header adapts as the viewport adjusts and I can adjust where the navigation is placed and for the hamburger to display on mobile, never, or always. If there are settings missing to accomplish the visual look, I encourage that we enhance the settings instead of trying to make a footer work like a header as that won't scale cleanly.
[d.pr/v/YVRO1m](d.pr/v/YVRO1m)

**Comment by pierorocca** [comment]
@yjailin as you look at improving and simplifying Blocks settings, this example of needing a slightly different header for mobile vs desktop is an interesting one to note.

**Comment by samueljseay** [comment]
> has the bug portion of this issue been separated out of this enhancement request?

@pierorocca Good point, I mentioned its a bug but there might be some contention if this is an enhancement or not because originally we had some code to try "our best" to handle issues if 2 were present on the page but iirc it wasn't a supported way to use the block.

> I'd like to get some design feedback on the original issue of mobile vs. desktop header differences and/or responsiveness. I'm not convinced that the workaround solution addresses the problem in the most optimal or performant way, and it likely complicates store editing

I 100% agree, the workarounds are clunky and not at all optimal.

**Comment by pierorocca** [comment]
@webmandesign Oliver I'm glad you found a workable workaround. Could you provide more context if or why a responsive header that adapts to the viewport width wouldn't work in your case?

**Comment by webmandesign** [comment]
@pierorocca In my theme I provide dedicated mobile header because it has different layout than the one that would be created by responsive scaling of desktop header.

It is a design decision for specific themes only and ultimately the user has the choice to use the double header (desktop + mobile) or single responsive header approach.

I spotted and reported this Mini Cart block issue as I've experienced it with this particular double header approach.

You can preview examples at:
- https://themedemos.webmandesign.eu/ileana/
- https://themedemos.webmandesign.eu/josephine/

**Comment by webmandesign** [comment]
**A bug update for WooCommerce 10.4:**

Now even the solution with only one Mini Cart block having `"addToCartBehaviour":"open_drawer"` enabled (see @samueljseay [comment above](#issuecomment-2357432848)) stopped working with WC10.4. No drawer is opened now when adding a product to cart when there are 2 Mini Cart blocks used in header.

Maybe it is related to [using interactivity API](https://developer.woocommerce.com/2025/12/10/woocommerce-10-4-the-interactivity-api-mini-cart-goes-live/#h-interactivity-api-mini-cart-is-now-the-default) now?

**Comment by SantosGuillamot** [comment]
@webmandesign Thanks for the report. I've been trying to replicate this, but having multiple minicarts in the header seems to work fine for me. The only issue I saw was an incorrect focus when closing it. I've opened [this pull request](https://github.com/woocommerce/woocommerce/issues/62493) for that. See this video (after the fix):

https://github.com/user-attachments/assets/49cd4f36-8b23-4bc1-9fac-44c25b47497a

Could you share more details about the steps to reproduce, please?

**Comment by webmandesign** [comment]
Hi @SantosGuillamot,

Thank you for the video.

The issue is related to _automatic drawer opening when adding a product to cart_. Allow me to explain in more details:

**In Site Editor:**

https://github.com/user-attachments/assets/a3df2d4d-b897-4323-8c2a-d5f942a271a2

1. Open "Header" template part for editing and add another Mini Cart block in there (so we have 2 of them).
2. Enable "Open drawer when adding" behavior for _**the first**_ Mini Cart block only.
3. Make sure the "Open drawer when adding" behavior is disabled for the second Mini Cart block.
4. Save your changes.

---

**On the front-end:**

https://github.com/user-attachments/assets/72456675-3fdc-40fc-86f4-c9321b621aed

1. In "Shop" page (anywhere the list of products is displayed) you can try if the drawer is opened when clicking any of the Mini Cart blocks. This should work.
2. Now add a product to the cart. No drawer is opened, even though we've set up one of the Mini Cart blocks to open the drawer.

---

**Presumption:**

I think the issue here is that WordPress interactivity API behaves according to the second (the latest) Mini Cart block setup.

If you switch the "Open drawer when adding" behavior setup for the Mini Cart blocks (→ disabling it for the first block, while enabling it for the second), the drawer opens when adding a product.
The same goes if you enable the drawer opening for both Mini Cart blocks.

---

**Expected behavior:**

I expect the drawer to be opened if any of the blocks has the option enabled.

**Comment by SantosGuillamot** [comment]
Thanks a lot for the detailed explanation 🙂 And sorry because I had missed some comments. If I understand the use cases correctly, and please correct me if I am wrong, you had different values for this setting as a workaround for a bug in the old version. Because it was opening multiple drawers and creating an unexpected user experience. Is that right? If that's not the case, could you clarify why you need the setting to have different values?

If I am not mistaken, the Interactivity API version should just load one drawer and work fine if that option is toggled on for all the minicarts you are adding. Is that not what you were expecting?

**Comment by webmandesign** [comment]
>  If I understand the use cases correctly, and please correct me if I am wrong, you had different values for this setting as a workaround for a bug in the old version. Because it was opening multiple drawers and creating an unexpected user experience. Is that right?

Yes, you are 100% correct.

> If I am not mistaken, the Interactivity API version should just load one drawer and work fine if that option is toggled on for all the minicarts you are adding. Is that not what you were expecting?

Indeed, the interactivity API loads just one drawer even if it is enabled for both blocks. So, this **original issue is actually resolved**.

I'm working on my themes updates where I enable drawers for both Mini Cart blocks, as now it is broken.
Though, as I create publicly available themes, users may have different needs and they may actually decide to disable the drawer for the mobile header Mini Cart. (Or they already have this disabled as they modified "Header" template part on their website, which is already saved in WP database, and my new theme update will not "fix" the drawer behavior for them.)

I still think this is an issue, although I understand it may be a new one. Should I open a new issue ticket for this?

**Comment by SantosGuillamot** [comment]
> I still think this is an issue, although I understand it may be a new one. Should I open a new issue ticket for this?

I would say it's probably better to create a new issue. It seems a complex problem because the "Open drawer when adding" happens outside of the minicart, so it is hard to know what is the expected behavior if there are two different settings for this. Maybe having a different discussion for that specific use case could be better.

**Comment by webmandesign** [comment]
@SantosGuillamot Reading your comment I had to stop and remember what my original issue actually was about (some time has past since I've logged it). And I realized I got confused after the current WC10.4 update when the drawer stopped working.

I was using the [aforementioned workaround](#issuecomment-2357432848) to make it work just like I wanted in my original issue description. But now I can simply enable the drawer for both Mini Cart blocks and it works.

The usecase from my previous latest comment does not make much sense :)

So, thank you, I'm closing this issue as it is resolved.

**Comment by glennbarrett** [comment]
I think the original issue here still exists. For example, I am using Kadence Theme with the advanced block headers, and I cannot add mini-cart to both the desktop and mobile layout of the header because it is only allowed to exist in one spot at a time. I have previously opened a ticket with Kadence on this, and they said it is due to the Woo limitation that was mentioned in the original issue comment here.

**Comment by webmandesign** [comment]
@glennbarrett I've solved this issue in my environment by modifying WC Mini Cart settings.

Here is a code you can use in your child theme's `functions.php` file:
```php
add_action( 'enqueue_block_editor_assets', function() {

	wp_register_script(
		'mini-cart-mods',
		'',
		[ 'wp-blocks', 'wp-hooks' ],
		false,
		true
	);

	wp_add_inline_script(
		'mini-cart-mods',
		"
		( () => {
			'use strict';

			wp.hooks.addFilter(
				'blocks.registerBlockType',
				'mini-cart-mods',
				( settings, name ) => {

					if ( 'woocommerce/mini-cart' === name ) {
						settings.supports.multiple = true;
					}

					return settings;
				},
				5
			);
		} )();
		"
	);

	wp_enqueue_script( 'mini-cart-mods' );
} );
```

It should enable multiple instances of Mini Cart block.

**Comment by luisherranz** [comment]
I'm going to check if there is still any impediment to allow the Mini Cart block to be inserted multiple times on the same page, or if it is a limitation that can be removed.
